### PR TITLE
Adversarial Training Setup

### PR DESCRIPTION
**Pull request by lucasrussell01**
Introducing the option to perform adversarial training by specifying input_type="Adversarial".
- Load an adversarial control dataset in tensorflow format and combining it with Shuffle Merge in DataLoader
- Introduce final dense layers + sigmoid output for data/MC discrimination in parallel to tau type classification
- k vector [k1; k2] controls the relative importance of tau type and data/MC classification in common layers where gradients defined as k1*classification_gradients - k2*adversarial_gradients
- Adversarial final layers optimised with an independent learning rate
- New metrics: adversarial loss (binary cross entropy or binary focal), binary adversarial_accuracy, average absolute value of gradients in common layers.
- Added option to restore weights and iterations of optimiser from previous model

### CODE REVIEW

File: Training/python/2018v1/Training_v0p1.py [hunk 1]
```diff
@@ -31,10 +31,13 @@
...
 from common import *
 import DataLoader
+import itertools
+import psutil
+
 
 class DeepTauModel(keras.Model):
 
-    def __init__(self, *args, loss=None, **kwargs):
+    def __init__(self, *args, loss=None, n_tau=None, use_AdvDataset=False, adv_parameter=[1,1], n_adv_tau=None, adv_learning_rate=None, **kwargs):
```
Comment (kandrosov) [inline Training/python/2018v1/Training_v0p1.py:40]: `n_tau = tf.shape(x)[0] - n_adv_tau` => no need to pass `n_tau` in the constructor.

File: Training/python/2018v1/Training_v0p1.py [hunk 2]
```diff
@@ -43,64 +46,131 @@ def __init__(self, *args, loss=None, **kwargs):
...
-        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
-        # Update metrics (including the ones that track losses)
-        self.loss_tracker.update_state(loss, sample_weight=sample_weight)
-        self.pure_loss_tracker.update_state(pure_loss, sample_weight=sample_weight) 
-        self.reg_loss_tracker.update_state(reg_loss)
-        self.compiled_metrics.update_state(y, y_pred, sample_weight)
+        # Forward Pass:
+        if self.use_AdvDataset:
+            with tf.GradientTape() as class_tape, tf.GradientTape() as adv_tape:
+                    y_pred = self(x, training=True)
```
Comment (kandrosov) [inline Training/python/2018v1/Training_v0p1.py:76]: do not duplicate the code inside `with` -> extract it in a sub-procedure.

File: Training/python/2018v1/Training_v0p1.py [hunk 3]
```diff
@@ -43,64 +46,131 @@ def __init__(self, *args, loss=None, **kwargs):
...
+        if self.use_AdvDataset:
+            class_layers = [var for var in self.trainable_variables if ("final" in var.name and "_adv" not in var.name)] # final classification dense only
+            adv_layers = [var for var in self.trainable_variables if ("final" in var.name and "_adv" in var.name)] #final adv only
+            common_layers = [var for var in self.trainable_variables if "final" not in var.name] #gradients common to both
+            grad_class = class_tape.gradient(loss, common_layers + class_layers) 
+            grad_adv = adv_tape.gradient(adv_loss, common_layers + adv_layers)
+            grad_class_excl = grad_class[len(common_layers):] # gradients of common part
+            grad_adv_excl = grad_adv[len(common_layers):] #gradients of adv part
+            grad_common = [self.k1*grad_class[i] - self.k2 * grad_adv[i] for i in range(len(common_layers))] 
+            mean_common = [tf.math.reduce_mean(tf.math.abs(grad_common[i])) for i in range(len(grad_common))]
```
Comment (kandrosov) [inline Training/python/2018v1/Training_v0p1.py:114]: - should be split into two metrics `mean_grad_class` and `mean_grad_common`
- I propose to add `tf.add_n(..) / len(grad_common)` to have a scalar value

File: Training/python/2018v1/Training_v0p1.py [hunk 4]
```diff
@@ -433,12 +517,36 @@ def run_training(model, data_loader, to_profile, log_suffix):
         input_shape, input_types = data_loader.get_input_config()
         data_train = tf.data.Dataset.from_generator(
             gen_train, output_types = input_types, output_shapes = input_shape
-            ).prefetch(tf.data.AUTOTUNE)
```
Comment (kandrosov) [inline Training/python/2018v1/Training_v0p1.py:436]: Keep `AUTOTUNE`

File: Training/python/2018v1/Training_v0p1.py [hunk 5]
```diff
@@ -433,12 +517,36 @@ def run_training(model, data_loader, to_profile, log_suffix):
...
+            ).prefetch(4)
         data_val = tf.data.Dataset.from_generator(
             gen_val, output_types = input_types, output_shapes = input_shape
-            ).prefetch(tf.data.AUTOTUNE)
```
Comment (kandrosov) [inline Training/python/2018v1/Training_v0p1.py:439]: keep `AUTOTUNE`

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Training/python/2018v1/Training_v0p1.py (hunk 3 above)
Comment (lucasrussell01) [reply Training/python/2018v1/Training_v0p1.py:114]: Assuming you meant mean_grad_class and mean_grad_adv, done

File: Training/python/2018v1/Training_v0p1.py [hunk 6]
```diff
@@ -366,12 +451,23 @@ def create_model(net_config, model_name, loss=None):
...
+                            kernel_initializer=dense_net_setup.kernel_init)(final_dense_adv)
+        sigmoid_output_adv = Activation("sigmoid", name="adv_output")(output_layer_adv)
+        model = DeepTauModel(input_layers, [softmax_output, sigmoid_output_adv], loss=loss, name=model_name, use_AdvDataset=True,
+                             adv_parameter=adv_param, n_adv_tau=n_adv_tau, adv_learning_rate=adv_learning_rate)
+    else:
+        model = DeepTauModel(input_layers, softmax_output, loss = loss, name=model_name)
     return model
 
-def compile_model(model, opt_name, learning_rate):
+def compile_model(model, opt_name, learning_rate, old_opt=None):
```
Comment (kandrosov) [inline Training/python/2018v1/Training_v0p1.py:467]: `old_opt` is not used in `compile_model`

**Review by kandrosov — CHANGES_REQUESTED** [review]

**Review by kandrosov — APPROVED** [review]
